curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -392,9 +412,66 @@ int FileCacheManager::HandleReadRequest(
         dataCache.first->AddReadDataCache(dataCache.second);
         VLOG(9) << "add read data cache end";
     }
+
     return 0;
 }
 
+void FileCacheManager::PrefetchS3Objs(std::vector<std::string> prefetchObjs) {
+    uint64_t blockSize = s3ClientAdaptor_->GetBlockSize();
+    uint32_t prefetchBlocks = s3ClientAdaptor_->GetPrefetchBlocks();
+    for (auto &obj : prefetchObjs) {
+        std::string name = obj;
+        // 加锁
+        downloadMtx_.lock();
+        if (downloadingObj_.find(name) != downloadingObj_.end()) {
+            LOG(INFO) << "whs--downloading exist: " << name
+                        << ", size: " << downloadingObj_.size();
+            downloadMtx_.unlock();
+            continue;
+        }
+        if (s3ClientAdaptor_->GetDiskCacheManager()->IsCached(name)) {
+            LOG(INFO) << "whs--downloading exist in cache: " << name
+                        << ", size: " << downloadingObj_.size();
+            downloadMtx_.unlock();
+            continue;
+        }
+        LOG(INFO) << "whs--download start all block: " << name
+                    << ", size: " << downloadingObj_.size();
+        downloadingObj_.emplace(name);
+        std::set<std::string>::iterator Iter;
+        Iter = downloadingObj_.begin();
+        // todo
+        for (; Iter != downloadingObj_.end(); Iter++) {
+            LOG(INFO) << "whs--download, downloading name = " << *Iter;
+        }
+        downloadMtx_.unlock();
+        auto task = [=]() {
+            char* dataCache_whs = new char[blockSize];  // +1?
+            if (dataCache_whs == nullptr) {
+                LOG(INFO) << "malloc failed, size = " << blockSize;
+            }
+            s3ClientAdaptor_->GetS3Client()->Download(
```

Comment:
This is a synchronization function， but it run in excution_queue and schedule by bthread, it will block and affect  brpc.

if excution_queue is used in user workspace, please make sure it can be finish quickly. Generally, we execute asynchronous calls in it.